coderd/database/dbrollup/dbrollup.go

```diff
@@ -0,0 +1,127 @@
+package dbrollup
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+)
+
+const (
+	// DefaultInterval is the default time between rollups.
+	// Rollups will be synchronized with the clock so that
+	// they happen 13:00, 13:05, 13:10, etc.
+	DefaultInterval = 5 * time.Minute
+)
+
+type Rolluper struct {
+	cancel context.CancelFunc
+	closed chan struct{}
+	db     database.Store
+	logger slog.Logger
+}
+
+// New creates a new DB rollup service that periodically runs rollup queries.
+// It is the caller's responsibility to call Close on the returned instance.
+//
+// This is for e.g. generating insights data (template_usage_stats) from
+// raw data (workspace_agent_stats, workspace_app_stats).
+func New(ctx context.Context, logger slog.Logger, db database.Store, interval time.Duration) *Rolluper {
+	ctx, cancel := context.WithCancel(ctx)
+
+	r := &Rolluper{
+		cancel: cancel,
+		closed: make(chan struct{}),
+		db:     db,
+		logger: logger.Named("dbrollup"),
+	}
+
+	//nolint:gocritic // The system rolls up database tables without user input.
+	ctx = dbauthz.AsSystemRestricted(ctx)
+	go r.start(ctx, interval)
+
+	return r
+}
+
+func (r *Rolluper) start(ctx context.Context, interval time.Duration) {
```

Comment:
suggestion: check if already closed

Reply:
Good catch, opted to not pass a context to New instead since we have Close.